integrations/access/servicenow/app.go

```diff
@@ -0,0 +1,621 @@
+		log.Error("ServiceNow plugin is not ready")
+	}
+
+	<-watcherJob.Done()
+
+	return trace.Wrap(watcherJob.Err())
+}
+
+func (a *App) init(ctx context.Context) error {
+	ctx, cancel := context.WithTimeout(ctx, initTimeout)
+	defer cancel()
+	log := logger.Get(ctx)
+
+	if validCred, err := credentials.CheckIfExpired(a.conf.Teleport.Credentials()); err != nil {
+		log.Warnf("Invalid Teleport credentials: %v", err)
+		if !validCred {
+			return trace.BadParameter(
+				"No valid credentials found, this likely means credentials are expired. In this case, please sign new credentials and increase their TTL if needed.",
+			)
+		}
+		log.Info("At least one non-expired credential has been found, continuing startup")
+	}
+
+	var err error
+
+	if a.teleport == nil {
+		bk := grpcbackoff.DefaultConfig
+		bk.MaxDelay = grpcBackoffMaxDelay
+		if a.teleport, err = client.New(ctx, client.Config{
+			Addrs:       a.conf.Teleport.GetAddrs(),
+			Credentials: a.conf.Teleport.Credentials(),
+			DialOpts: []grpc.DialOption{
+				grpc.WithConnectParams(grpc.ConnectParams{Backoff: bk, MinConnectTimeout: initTimeout}),
+				grpc.WithReturnConnectionError(),
+			},
+		}); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+
+	pong, err := a.checkTeleportVersion(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	webProxyURL, err := url.Parse(pong.ProxyPublicAddr)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	a.conf.ClientConfig.WebProxyURL = webProxyURL
+	a.serviceNow, err = NewClient(a.conf.ClientConfig)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	return nil
+}
+
+func (a *App) checkTeleportVersion(ctx context.Context) (proto.PingResponse, error) {
+	log := logger.Get(ctx)
+	log.Debug("Checking Teleport server version")
+
+	pong, err := a.teleport.Ping(ctx)
+	if err != nil {
+		if trace.IsNotImplemented(err) {
+			return pong, trace.Wrap(err, "server version must be at least %s", minServerVersion)
+		}
+		log.Error("Unable to get Teleport server version")
+		return pong, trace.Wrap(err)
+	}
+	err = lib.AssertServerVersion(pong, minServerVersion)
+	return pong, trace.Wrap(err)
+}
+
+func (a *App) onWatcherEvent(ctx context.Context, event types.Event) error {
+	if kind := event.Resource.GetKind(); kind != types.KindAccessRequest {
+		return trace.Errorf("unexpected kind %s", kind)
+	}
+	op := event.Type
+	reqID := event.Resource.GetName()
+	ctx, _ = logger.WithField(ctx, "request_id", reqID)
+
+	switch op {
+	case types.OpPut:
+		ctx, _ = logger.WithField(ctx, "request_op", "put")
+		req, ok := event.Resource.(types.AccessRequest)
+		if !ok {
+			return trace.Errorf("unexpected resource type %T", event.Resource)
+		}
+		ctx, log := logger.WithField(ctx, "request_state", req.GetState().String())
+
+		var err error
+		switch {
+		case req.GetState().IsPending():
+			err = a.onPendingRequest(ctx, req)
+		case req.GetState().IsApproved():
+			err = a.onResolvedRequest(ctx, req)
+		case req.GetState().IsDenied():
+			err = a.onResolvedRequest(ctx, req)
```

Comment:
Can we replace the `IsApproved` and `IsDenied` (and merge the case branch) with `IsResolved`. Should be the same thing for now.

There's an ongoing PR that adds a new state.
https://github.com/gravitational/teleport/pull/31346/files
This way seems safer: less probability of someone forgetting to consider the new state